test/implementations/functions_testing.py

```diff
@@ -8,436 +7,659 @@
+        # Expected unitary after calculation is identity.
+        expected_unitary = identity_matrix(no_of_qubits)
 
+        # X_1 * X_2^dagger * np.conj((X_1 * X_2^dagger)[0,0]) = I
+        m = np.matmul(b, inverse_matrix)
+        generated_unitary = m * np.conjugate(m[0, 0])
 
-# 5.2 Dirty Wasted Relative-phase (entangled left-out)
-# In here, similarly as for the clean-wasted case, this class is reducible to
-# the regular MCT entangled left-out.
+        if generated_unitary.shape != expected_unitary.shape:
+            return False, "Unitary B has different shape."
 
-# 5.3 Dirty Wasted seperable left-out
+        if (
+            np.allclose(
+                generated_unitary,
+                expected_unitary,
+                atol=absolute_error_tol,
+                rtol=relative_error_tol,
+            )
+            is False
+        ):
+            return False, "Something wrong with the implementation"
+
+        return True, ""
+    except Exception:
+        return False, ""
+
+
+# 5.4 Dirty Wasted Relative-phase seperable left-out
+def verify_circuit_dirty_wasted_relative_separable_auxiliary(
+    unitary_matrix, controls_no: int, auxiliaries_no: int
+):
+    if auxiliaries_no == 0:
+        return False, "No of Auxiliary qubit should bigger than 0"
+    try:
+        # get mct inverse matrix
+        inverse_matrix = load_matrix("reverse_noauxiliary", controls_no)
+
+        b_shape = 2 ** (controls_no + 1)
+        c_shape = 2 ** (auxiliaries_no)
+
+        b, c = reverse_kronecker_product(unitary_matrix, (b_shape, b_shape))
+
+        no_of_qubits = controls_no + 1
+
+        # check shape of c
+        if c.shape != (c_shape, c_shape):
+            return False, "Unitary C has different shape."
+
+        # check if c is unitary
+        check_c = np.matmul(c, np.linalg.inv(c))
+
+        if (
+            np.allclose(
+                check_c,
+                identity_matrix(auxiliaries_no),
+                atol=absolute_error_tol,
+                rtol=relative_error_tol,
+            )
+            is False
+        ):
+            return False, "Something wrong with the implementation"
+
+        # Expected unitary after calculation is identity.
+        expected_unitary = identity_matrix(no_of_qubits)
+
+        # X_1 * X_2^dagger = I
+        m = np.matmul(b, inverse_matrix)
+        generated_unitary = np.absolute(m)
+
+        if generated_unitary.shape != expected_unitary.shape:
+            return False, "Unitary B has different shape."
+
+        if (
+            np.allclose(
+                generated_unitary,
+                expected_unitary,
+                atol=absolute_error_tol,
+                rtol=relative_error_tol,
+            )
+            is False
+        ):
+            return False, "Something wrong with the implementation"
+
+        return True, ""
+    except Exception:
+        return False, ""
+
+
+def verify_dependencies(rd):
+    if rd["DNW"]:
+        assert rd["CNW"]

```

Comment:
I would say this should run recursively `verify_dependencies` on `CNW`, etc.

Reply:
this is optional for this PR. But could be next thing to do

Reply:
i was thinking to put in the functions_testing, but it seems this one is not working. I forgot to remove this

Reply:
i will add for recursive as an issue

Reply:
moved to issue #15